libc-top-half/musl/src/string/strspn_simd.c

```diff
@@ -0,0 +1,175 @@
+
+  v128_t indices_0_7 = v & wasm_u8x16_const_splat(0x8f);
+  v128_t indices_8_15 = indices_0_7 ^ wasm_u8x16_const_splat(0x80);
+
+  v128_t row_0_7 = wasm_i8x16_swizzle(bitmap.lo, indices_0_7);
+  v128_t row_8_15 = wasm_i8x16_swizzle(bitmap.hi, indices_8_15);
+
+  v128_t bitsets = row_0_7 | row_8_15;
+  return wasm_i8x16_eq(bitsets & bitmask, bitmask);
+}
+
+size_t strspn(const char *s, const char *c)
+{
+  // Note that reading before/after the allocation of a pointer is UB in
+  // C, so inline assembly is used to generate the exact machine
+  // instruction we want with opaque semantics to the compiler to avoid
+  // the UB.
+  uintptr_t align = (uintptr_t)s % sizeof(v128_t);
+  uintptr_t addr = (uintptr_t)s - align;
+
+  if (!c[0]) return 0;
+  if (!c[1]) {
+    v128_t vc = wasm_i8x16_splat(*c);
+    for (;;) {
+      v128_t v;
+      __asm__(
+          "local.get %1\n"
+          "v128.load 0\n"
+          "local.set %0\n"
+          : "=r"(v)
+          : "r"(addr)
+          : "memory");
+      v128_t cmp = wasm_i8x16_eq(v, vc);
+      // Bitmask is slow on AArch64, all_true is much faster.
+      if (!wasm_i8x16_all_true(cmp)) {
+        // Clear the bits corresponding to align (little-endian)
+        // so we can count trailing zeros.
+        int mask = (uint16_t)~wasm_i8x16_bitmask(cmp) >> align << align;
+        // At least one bit will be set, unless align cleared them.
+        // Knowing this helps the compiler if it unrolls the loop.
+        __builtin_assume(mask || align);
+        // If the mask became zero because of align,
+        // it's as if we didn't find anything.
+        if (mask) {
+          // Find the offset of the first one bit (little-endian).
+          return addr - (uintptr_t)s + __builtin_ctz(mask);
+        }
+      }
+      align = 0;
+      addr += sizeof(v128_t);
+    }
+  }
+
+  __wasm_v128_bitmap256_t bitmap = {};
+
+  for (; *c; c++) {
+    // Terminator IS NOT on the bitmap.
+    __wasm_v128_setbit(&bitmap, *c);
+  }
+
+  for (;;) {
+    v128_t v;
+    __asm__(
+        "local.get %1\n"
+        "v128.load 0\n"
+        "local.set %0\n"
+        : "=r"(v)
+        : "r"(addr)
+        : "memory");
+    v128_t cmp = __wasm_v128_chkbits(bitmap, v);
+    // Bitmask is slow on AArch64, all_true is much faster.
+    if (!wasm_i8x16_all_true(cmp)) {
+      // Clear the bits corresponding to align (little-endian)
+      // so we can count trailing zeros.
+      int mask = (uint16_t)~wasm_i8x16_bitmask(cmp) >> align << align;
+      // At least one bit will be set, unless align cleared them.
+      // Knowing this helps the compiler if it unrolls the loop.
+      __builtin_assume(mask || align);
+      // If the mask became zero because of align,
+      // it's as if we didn't find anything.
+      if (mask) {
+        // Find the offset of the first one bit (little-endian).
+        return addr - (uintptr_t)s + __builtin_ctz(mask);
+      }
+    }
```

Comment:
We keep on using this pattern; is the logic repetitive enough that we should define a helper function somewhere?

Reply:
Yeah, but it's always subtly different.

Some times it's (1) `cmp`, `any_true` and `bitmask`; others (2) `cmp`, `all_true` and `(uint16_t)~bitmask`.

Then, `strlen` gets away with doing `!all_true` before the `cmp` and `bitmask`; `memrchr` does `clz` instead of `ctz`.

I also don't see how to get those two main patterns to inline without going into macros, and inlining is important. That `__builtin_assume` makes the compiler figure out it's good to unroll the first iteration to make the subsequent ones significantly simpler.